### PR TITLE
Error when [foo.(bar)] is used with nonprojection [bar], warn if [bar] nonprimitive projection.

### PR DESCRIPTION
**Pull request by SkySkimmer**
Maybe it would make more sense to turn some of the stdlib records into primitive records rather than stop using the primitive projection syntax. Thoughts?

cf https://gitter.im/coq/coq?at=5bd31d8b600c5f64230f50a2

Overlay https://gitlab.mpi-sws.org/iris/lambda-rust/merge_requests/13

### CODE REVIEW

**Comment by herbelin** [comment]
> Maybe it would make more sense to turn some of the stdlib records into primitive records rather than stop using the primitive projection syntax. Thoughts?

The meaning of "primitive projections" should be first defined. Strictly speaking, I don't see how to  define "primitive projections" differently than what @mattam82 initially did, i.e. as a compact representation of projections. Then, accepting this definition (rather than defining it as attached to "negative" types, i.e. with forbidden `match`), we don't have to wait to represent all projections (even the ones for positive types) in a compact way (up to the incompatibility in referring to subterms of the then-hidden parameters of the projections). By doing so, projections would become a synonymous of compactly-represented projections, as well as a synonymous of "native" projections.

In particular, I approve that we forbid being able to write `0.(S)` to mean `S 0`, as the system allows currently, but I would say that the warning should only be when the "ref" is not a "projection" (primitive or not in a first time, but without having to make a difference in the longer term when all full applications to a projection are eventually represented in a compact way).

**Comment by SkySkimmer** [comment]
I prefer to have the syntax correspond to the internal representation.

**Comment by ppedrot** [comment]
See also #8838, we might want to allow `t.(0)` to anonymously access the first field of a record.

**Comment by herbelin** [comment]
> I prefer to have the syntax correspond to the internal representation.

Note, that, historically, the dot syntax was for `Record`'s projections, independently of the existence of a primitive internal representation of projections.

But, again, if we accept that all projections deserve a compact representation (independently of whether we think at this or that inductive type as `match`-based or projection-based, since, after all, `match`-based types can also be equipped with projections), this shall exactly put the superficial syntax in correspondence with the internal use of a `Constr.Proj` constructor.

> See also #8838, we might want to allow t.(0) to anonymously access the first field of a record.

What would be the advantages of not using a name? Will you skip the named fields in the numbering or will it be an alternative way to name fields?

**Comment by ppedrot** [comment]
> will it be an alternative way to name fields?

Rather that option.

**Review by ppedrot — APPROVED** [review]
FTR, I like this.

**Comment by ppedrot** [comment]
I think that this deprecation should be thoroughly discussed.

I am personally in favour of it, and if I read @herbelin's comment correctly he is not against, as this is somewhat orthogonal to the primitive projection implementation, as long as there is an actual notion of projection involved.

What about others? @mattam82 is the original designer of the primitive projection system so he should be involved in the discussion. @maximedenes has been suffering with them during the Great Notation Recession, and I think he would be happy if we could clean up the syntactic sugar.

This is probably also interesting for the few users of primitive projections out there, e.g. @RalfJung and @robbertkrebbers.

**Comment by mattam82** [comment]
I like this too. However Hugo, I’m not sure it’s so easy to support the
notion that Proj(p,c) applies to any record, if you still allow `match` on
the same inductive as you seem to propose (while ideally yes, it would be
cool to allow compact representations as much as possible). For example,
you would have to unfold the primitive application to a `match` too, to
avoid introducing a morally equivalent neutral term. Unfolding the
primitive application to a match would require typing information to
retrieve the parameters, making reduction depend on type inference. This
problematic issue is relegated to the compatibility layer right now, that’s
basically what expand_projection does, and it’s not nice to look at.

Maybe you have another option in mind?
— Matthieu

**Comment by SkySkimmer** [comment]
>For example, you would have to unfold the primitive application to a `match` too, to avoid introducing a morally equivalent neutral term.

How important is it to avoid having a morally equivalent neutral term?

**Comment by herbelin** [comment]
In OCaml, tuples comes with both primitive destructing let and primitive projections (e.g. the code of `Pervasives.fst` is `external fst : 'a * 'b -> 'a = "%field0"`). Why couldn't we do the same?

In particular, I don't see an obstacle to have one reduction rule for `let (x,y) := (t,u) in c` and another distinct one for `fst (t,u)`. We shall want to relate the two forms, but this can be done in the conversion. There would be a priori different ways to do that, but, one of them could be to reduce (in the conversion) `let (x,y) := t in u` to `u[x:=fst t][y:=snd t]`. We already provide definitional eta on non-recursive non-unit primitive projections and this reduction does not seem to require more than that (details to check but I'm pretty confident).

Having both `match` and projections in the same type would help a lot for the compatibility since one of the problems for a straightforward move of the standard library to primitive projections is that the `match` becomes syntactic sugar and stops reducing the same. After all, from a computational point of view, both `match` and projections have their interest. The `match` is for instance able to provide a sharing blocking on neutral terms in situations where expanding it to `fst` and `snd` would increase the size of the term.

One may then wonder what becomes the notion of unique normal form and indeed this is what we would loose: reduction tactics would give different normal forms for terms intended to be equal. But after all, this is already the case with eta since eta is not considered in reduction tactics. And, similarly, we could justify these different normal forms by saying that they differ up to eta. Or we could also provide new reduction tactics which apply on demand eta-reduction, or eta-contraction, or hence also eta-expansion/reduction for match, or eta-expansion/reduction for record.

**Comment by mattam82** [comment]
Indeed, if you go the other way and reduce match on a neutral to a substitution of projections then you can equate the let and projection variants, at the cost of changing the complexity but without depending on type inference. This should be fine even for recursive records. 

Just thinking out loud here, but there is also the case of records with non-definable projections due to sort restriction. If we expanded matches on those to projections we'd have to use "Anonymous" projection somehow? For example:
```
Record myRec : Prop := Rec_intro { somenat : nat; someprop : True }.
```
```
match t : myRec return P : Prop with Rec_intro x y => b
```
would reduce to
``` 
b[t.0/x, t.1/y]
```

**Comment by SkySkimmer** [comment]
>Indeed, if you go the other way and reduce match on a neutral to a substitution of projections then you can equate the let and projection variants, at the cost of changing the complexity but without depending on type inference. This should be fine even for recursive records.

I think the idea is rather to have the conversion `match foo with {| bla = x |} => x end == foo.(bla)` without changing the reduction.
Then nondefinable projections don't come up.

**Comment by mattam82** [comment]
I meant implementing that reduction during conversion, i.e. when finding a neutral term `n` of a record type as a discriminee of a match, expand it to `mkR r.0 ... r.n` and continue conversion on the weak-head normal forms.

**Comment by SkySkimmer** [comment]
So in conclusion do we relax the warning or merge as-is?

**Comment by herbelin** [comment]
> Just thinking out loud here, but there is also the case of records with non-definable projections due to sort restriction. 

>  Then nondefinable projections don't come up.

Right. For these types, surjective pairing does not hold and the only eta-conversion which we can think of is `E[let (...) := ... in ...] == let (...) := ... in E[...] end` for `E` with input and output in the allowed sorts. That seems to be less easy to deal with in the conversion. However, thinking aloud too, even if replacing `let` with `projections` and locally relying on untyped terms, I doubt that reducing `let` into projections could lead to wrong conversion problems in the sense that I conjecture that for any conversion problem going through such locally ill-typed terms, there is a conversion path which is well-typed.

> So in conclusion do we relax the warning or merge as-is?

My reconstruction of @mattam82 and @ppedrot's view (though I'm not 100% sure and it would be better that they express themselves) is that they don't care whether it is for primitive projections only or for all projections. Personally, I'm afraid that a strict warning causes more noise than needed in existing developments, even more if all kinds of projections are eventually turned into compactly represented projections. On the other side, the warning can help to identify what needs to be done in the long term. So, maybe it depends on your own plans?

**Comment by SkySkimmer** [comment]
Maybe what we should do is error when it's not any kind of projection and warn when it's an emulated projection. Then developments which are OK with using `.(foo)` with emulated projections can disable the warning but we don't see total nonsense like `0.(S)`.

**Comment by ppedrot** [comment]
We could probably do even better, i.e. by erroring on non-projections we can still bring the non-primitive and primitive projections closer, by saying that `t.(p)` is restricted to `p` that are registered projections, and forbid the presence of parameters there. Registering non-primitive projection is somewhat already done for other purposes, we could leverage this.

**Comment by SkySkimmer** [comment]
Updated

**Comment by SkySkimmer** [comment]
How do I make a PR on lambda-rust?

**Comment by ejgallego** [comment]
> How do I make a PR on lambda-rust?

You need to go to their gitlab I think @RalfJung ?

**Comment by RalfJung** [comment]
Yes, please create an account at https://gitlab.mpi-sws.org/users/sign_in and then tell me your username so I can unlock your account to create forks.

**Comment by SkySkimmer** [comment]
@RalfJung https://gitlab.mpi-sws.org/SkySkimmer

**Comment by RalfJung** [comment]
@SkySkimmer you can create forks now.

**Comment by SkySkimmer** [comment]
@Zimmi48 Multiple spurious-looking failures in https://gitlab.com/coq/coq/pipelines/48021781 if you want to have a look (lambda-rust failure isn't spurious, I had a bad overlay)

**Comment by SkySkimmer** [comment]
In https://gitlab.mpi-sws.org/iris/lambda-rust/merge_requests/13#note_33219 @RalfJung said 
>Hm, I have to admit I quite liked being able to use "fake projections", it communicates intent very well IMO. But I suppose you have good reasons for removing that ability.

So this syntax was to some extent deliberately used with non-projections.

**Comment by RalfJung** [comment]
Yes, it was certainly deliberate. We were thinking of it as a "user-defined" projection, and at least I was quite happy to be able to use this syntax. I don't recall if @jhjourdan opposed or agreed. ;)

**Comment by mattam82** [comment]
In an ideal situation compact projections could use the r.p syntax leaving the r.(f) syntax free... That requires disambiguating this from a module qualification at interpretation time but I don’t see why we couldn’t do that.

**Comment by ppedrot** [comment]
> I don’t see why we couldn’t do that

Because then you wouldn't be able to use qualified name for projections, as there would be an ambiguity. This looks like a very bad idea to me.

**Comment by mattam82** [comment]
Thanks, I had forgotten about the ambiguity.

**Comment by ppedrot** [comment]
I am assigning this to myself because I think this is a preliminary but critical part of cleaning up the mess around the compatibility layer for primitive projections. In particular, I am strongly advocating for this to be part of 8.10, otherwise we'll just delay fixing the issues to a later time again.

My ultimatum mood leads me to believe that there is no opposition so far to this PR, and thus that I can safely propose a merge by next week if nobody opposes it more fiercely. The patch in itself is quite small, if more people want to review they're welcome.

**Comment by mattam82** [comment]
Can we turn the warning off by default (leaving the error for nonprojections), until we have a clear path for moving everyone to a compact representation? I like the goal but I don't want to spam users like this until we have a beneficial alternative to offer.

**Comment by ppedrot** [comment]
@SkySkimmer would you mind tweaking your PR? Thanks.

**Comment by SkySkimmer** [comment]
changed

Gaëtan Gilbert

**Comment by ppedrot** [comment]
Will merge this tomorrow if no opposition.

**Comment by ppedrot** [comment]
Erf, forgot about this one. @SkySkimmer can you rebase so that this lands in 8.10?

**Comment by SkySkimmer** [comment]
rebased

Gaëtan Gilbert

**Comment by ppedrot** [comment]
Mtac2 broken: consequence or coincidence?

**Comment by SkySkimmer** [comment]
It fails on master too https://gitlab.com/coq/coq/-/jobs/187407939

**Comment by JasonGross** [comment]
Bedrock2 needs an overlay / patch; this was missed because we were previously masking errors via a pipe.  C.f. https://github.com/coq/coq/pull/9872#issuecomment-478534705

**Comment by tchajed** [comment]
Argosy was actually broken by this PR, but it was added to the CI a day late to break the build (everything is fixed upstream now).

**Comment by ejgallego** [comment]
Thanks @JasonGross @tchajed

**Comment by cpitclaudel** [comment]
This patch broke our code too when updating to 8.10.  If the syntax restriction that this introduces is just a matter of taste, can we revert it and discuss it on somewhere more visible than the github tracker? The discussion mentions primitive projections a lot, but what bit us was the new syntactic restriction, and I don't get a clear picture of the connection between both from this discussion. Sorry if I miss something.

AFAICT, @RalfJung and @jhjourdan both expressed support for the old feature when merging the overlay, and this broke a bunch of other projects.  I too find the feature quite useful for writing more readable terms.

Additionally, what is part of a typeclass definition and what is a function derived from it is arguably an implementation detail when you're using the typeclass (as opposed to defining instances), so the previous syntax was quite convenient (and it provided a nice way to do backwards-compatible changes to records and typeclasses without forcing clients to update their syntax).

**Comment by tchajed** [comment]
For the record, I also liked using the projection syntax for functions the way we were in Argosy. I agree with @cpitclaudel that it's an implementation detail whether something is part of a typeclass or is derived.

**Comment by samuelgruetter** [comment]
Being able to use projection syntax for functions would be nice indeed, especially once https://github.com/coq/coq/issues/9814 is fixed.
How about this: Add a switch to the `Arguments` vernac to control whether a function uses projection syntax, which would apply to both parsing and printing, rather than hardcoding this criterion to be whether it's a primitive projection.
